SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,710 @@
+  /// %result = partial_apply %takesClosure(%reabstractedClosure) : $@convention(thin) (@owned @callee_guaranteed () -> @out Int) -> Int
+  /// ret %result
+  /// ```
+  var convertedAndReabstractedClosures = InstructionSet(context)
+
+  defer {
+    convertedAndReabstractedClosures.deinitialize()
+  }
+
+  var callSiteMap = CallSiteMap()
+
+  for inst in caller.instructions {
+    if !convertedAndReabstractedClosures.contains(inst),
+       let rootClosure = inst.asSupportedClosure
+    {
+      updateCallSites(for: rootClosure, callSiteMap: &callSiteMap, 
+                      convertedAndReabstractedClosures: &convertedAndReabstractedClosures, context)
+    }
+  }
+
+  return callSiteMap.callSites
+}
+
+// ===================== Utility types, extensions and functions ===================== //
+
+private struct OrderedDict<Key: Hashable, Value> {
+  private var valueIndexDict: [Key: Int] = [:]
+  private var entryList: [(Key, Value)] = []
+
+  public subscript(key: Key) -> Value? {
+    if let index = valueIndexDict[key] {
+      return entryList[index].1
+    }
+    return nil
+  }
+
+  public mutating func insert(key: Key, value: Value) {
+    if valueIndexDict[key] == nil {
+      valueIndexDict[key] = entryList.count
+      entryList.append((key, value))
+    }
+  }
+
+  public mutating func update(key: Key, value: Value) {
+    if let index = valueIndexDict[key] {
+      entryList[index].1 = value
+    }
+  }
+
+  public var keys: LazyMapSequence<Array<(Key, Value)>, Key> {
+    entryList.lazy.map { $0.0 }
+  }
+
+  public var values: LazyMapSequence<Array<(Key, Value)>, Value> {
+    entryList.lazy.map { $0.1 }
+  }
+}
+
+private typealias CallSiteMap = OrderedDict<PartialApplyInst, CallSite>
+
+private extension CallSiteMap {
+  var callSites: [CallSite] {
+    Array(self.values)
+  }
+}
+
+private extension PartialApplyInst {
+  /// True, if the closure obtained from this partial_apply is the
+  /// pullback returned from an autodiff VJP
+  var isPullbackInResultOfAutodiffVJP: Bool {
+    if self.parentFunction.isAutodiffVJP,
+       let use = self.uses.singleUse,
+       let tupleInst = use.instruction as? TupleInst,
+       let returnInst = self.parentFunction.returnInstruction,
+      tupleInst == returnInst.returnedValue
+    {
+      return true
+    }
+
+    return false
+  }
+
+  var hasNonInoutIndirectArguments: Bool {
+    self.argumentOperands
+      .filter { !$0.value.type.isObject }
+      .allSatisfy {
+        if let convention = self.convention(of: $0) {
+          return convention == .indirectInout || convention == .indirectInoutAliasable
+        }
+        return false
+      } 
+  }
+}
+
+private extension Instruction {
+  var asSupportedClosure: SingleValueInstruction? {
+    switch self {
+    case let tttf as ThinToThickFunctionInst where tttf.callee is FunctionRefInst:
+      return tttf
+    case let pai as PartialApplyInst where pai.callee is FunctionRefInst && pai.hasNonInoutIndirectArguments:
```

Comment:
It's required that the closure only have inout-indirect arguments. However, I'm not sure why. 

It's something I've tried to figure out via the commits/PRs but I've usually run into a dead end. Here's the [commit](https://github.com/apple/swift/commit/597dc8ef645b1dbdeb168dbcc115470539bab027) where the change to support any indirect arguments at all was added.

@atrick seems to have last touched that code [here](https://github.com/apple/swift/commit/30e3062bf25b03a4238dc68bee568db8652f5700), although the actual change seems rather tangential to why certain indirect types aren't supported.

I have created a swift forums post [here](https://forums.swift.org/t/non-inout-indirect-types-not-supported-in-closure-specialization-optimization/70826) asking why non-inout indirect arguments aren't supported in the closure-specialization optimization. Perhaps you can engage someone who might have an answer?

Reply:
Okay, then let's add a TODO.
Non-inout indirect arguments shouldn't be that important, because after generic specialization (most) arguments are passed directly.

Reply:
It's not completely clear to me what the TODO should be. Do we want to handle non-inout indirect arguments then?

I also did not fully understand Joe's reply on the Swift forums post.
> There shouldn't be any fundamental reason. It may not have been a priority because the "opaque values" change that would push address lowering for indirect types to late in the pipeline has been coming "real soon now" for a while, but it would still be valuable to implement in the mean time.

Would you be able to briefly explain what the "opaque values" change is and how/why it affects indirect arguments?

Additionally, are non-inout arguments affected differently from inout arguments by this change?

Reply:
"opaque values" means that we don't have address-only types in SIL. So the only indirect parameters would be inout arguments.

Let's add a TODO, like "TODO: figure out what to do with non-inout indirect arguments"

Reply:
> Non-inout indirect arguments shouldn't be that important, because after generic specialization (most) arguments are passed directly.

As far as I saw from lots of autodiff-related code it's not quite so. Either something is preventing the specialization, or... I do not know :) Anyway, it would be great to have some generic-related tests as we need to ensure that these cases are handled properly.